File: refactored/ruby/gpt_api.rb
require 'net/http'
require 'json'

def gpt_api(prompt, model: 'gpt-3.5-turbo', temperature: 0.5, max_tokens: 50,
            system_message: nil, num_retries: 3, pause_base: 1,
            presence_penalty: 0.0, frequency_penalty: 0.0)

  messages = [{ 'role' => 'user', 'content' => prompt }]
  messages.unshift({ 'role' => 'system', 'content' => system_message }) if system_message

  url = URI.parse('https://api.openai.com/v1/chat/completions')
  headers = { 'Content-Type' => 'application/json',
              'Authorization' => "Bearer #{ENV['OPENAI_API_KEY']}" }

  response = nil
  num_retries.times do |retry_count|
    http = Net::HTTP.new(url.host, url.port)
    http.use_ssl = true

    request = Net::HTTP::Post.new(url.path, headers)
    request.body = {
      'model' => model.to_s,
      'temperature' => temperature,
      'max_tokens' => max_tokens,
      'messages' => messages,
      'presence_penalty' => presence_penalty,
      'frequency_penalty' => frequency_penalty
    }.to_json

    response = http.request(request)
    break if response.is_a?(Net::HTTPSuccess)
    sleep pause_base * (retry_count + 1)
  end

  unless response.is_a?(Net::HTTPSuccess)
    puts "Failed to make a successful API request. Error: #{response.body}"
    return nil
  end

  choices = JSON.parse(response.body)['choices']
  if choices.length > 0
    message = choices[0]['message']['content']
  else
    message = 'The model did not return a message. You may need to increase max_tokens.'
  end

  clean_message = message.gsub("\n", ' ') # replace newlines with spaces
  clean_message.strip! # trim leading and trailing white spaces
  clean_message
end

File: refactored/python/gpt_api.py
import os
import requests
import json
import time
from tenacity import retry, stop_after_attempt, wait_fixed

# Set API key as system variable
def set_api_key(api_key):
    os.environ["OPENAI_API_KEY"] = api_key

set_api_key("sk-paste-your-api-key-here")
print(os.environ["OPENAI_API_KEY"])

def create_headers():
    return {
        "Authorization": f"Bearer {os.getenv('OPENAI_API_KEY')}",
        "Content-Type": "application/json"
    }

def create_data(prompt, model, temperature, max_tokens, messages, presence_penalty, frequency_penalty):
    return {
        "model": model,
        "temperature": temperature,
        "max_tokens": max_tokens,
        "messages": messages,
        "presence_penalty": presence_penalty,
        "frequency_penalty": frequency_penalty
    }

def handle_response(response):
    if len(response['choices']) > 0:
        message = response['choices'][0]['message']['content']
    else:
        message = "The model did not return a message. You may need to increase max_tokens."
    return message.replace("\n", " ").strip()

@retry(stop=stop_after_attempt(3), wait=wait_fixed(1))
def send_request(headers, data):
    response = requests.post("https://api.openai.com/v1/chat/completions", headers=headers, data=json.dumps(data))
    response.raise_for_status()
    return response.json()

def gpt_api(prompt, model="gpt-3.5-turbo", temperature=0.5, max_tokens=50, 
            system_message=None, presence_penalty=0.0, frequency_penalty=0.0):

    messages = [{"role": "user", "content": prompt}]
    if system_message is not None:
        messages.insert(0, {"role": "system", "content": system_message})
    
    headers = create_headers()
    data = create_data(prompt, model, temperature, max_tokens, messages, presence_penalty, frequency_penalty)

    response_json = send_request(headers, data)
    
    return handle_response(response_json)
